### PR TITLE
Add writev to TcpStream

### PR DESCRIPTION
**Pull request by fasterthanlime**
Seemed like a glaring omission. Not sure how this fits in with in-progress work, let's see if this alone works.

### CODE REVIEW

File: src/driver/socket.rs [hunk 1]
```diff
@@ -44,6 +44,11 @@ impl Socket {
         op.write().await
     }
 
+    pub async fn writev<T: IoBuf>(&self, buf: Vec<T>) -> crate::BufResult<usize, Vec<T>> {
```
Comment (ollie-etl) [inline src/driver/socket.rs:47]: Could this be generalised to 

```suggestion
    pub async fn writev<T: IoBuf, B: Deref<Target = [T]>>(&self, buf: Buf) -> crate::BufResult<usize, Buf> {
```

Comment (fasterthanlime) [reply src/driver/socket.rs:47]: I'm not really fond of the `Vec<T>` requirement either here or internally:

https://github.com/tokio-rs/tokio-uring/blob/913aa14c567cfae026b7d4f88a5155921cfb9575/src/driver/writev.rs#L22

And I am using `smallvec` in a bunch of places but I'm always wary of stuff moving after it's been sent to the uring submission queue. I guess this small change is probably harmless though? I wonder what @Noah-Kennedy thinks.

Comment (fasterthanlime) [reply src/driver/socket.rs:47]: Ah, we'd also need to make the Future type generic over the buffer type.

**Comment by Noah-Kennedy** [comment]
@fasterthanlime this would be nice to have in the crate.

**Comment by fasterthanlime** [comment]
Note that in `tokio`, the equivalent method is named `write_vectored`: https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_vectored

I also haven't written tests for it yet.

**Comment by Noah-Kennedy** [comment]
I'm down to merge once tests are in. I'm fine with the Vec dependency here.

File: src/driver/socket.rs (hunk 1 above)
Comment (FrankReh) [reply src/driver/socket.rs:47]: Making the op futures generic on the buffer type seems to be the direction in general. The logic around op futures was simplified with the Completable trait recently and the Completable trait already is generic on the buffer type. .. If I understand your discussion here correctly ..

Comment (FrankReh) [reply src/driver/socket.rs:47]: It seems fine to be following what was done for the File.writev_at at this time.

File: src/net/tcp/stream.rs [hunk 2]
```diff
@@ -149,6 +149,34 @@ impl TcpStream {
...
+    /// The method returns the operation result and the same array of buffers
+    /// passed in as an argument. A return value of `0` typically means that the
+    /// underlying socket is no longer able to accept bytes and will likely not
+    /// be able to in the future as well, or that the buffer provided is empty.
+    ///
+    /// # Errors
+    ///
+    /// Each call to `write` may generate an I/O error indicating that the
+    /// operation could not be completed. If an error is returned then no bytes
+    /// in the buffer were written to this writer.
```
Comment (FrankReh) [inline src/net/tcp/stream.rs:170]: This comment about no bytes in the buffer seems suspect. The writev is of a list of buffers, the comment makes it sound like a single buffer. I don't see a kernel man page for writev making this kind of statement. But this comment was already made for the File writev_at function so no need to change it here. Just sayin, it looks odd.

**Review by FrankReh — CHANGES_REQUESTED** [review]
These changes look good. Would you mind adding the same function for the UnixStream?

And if you wouldn't mind merging master in again. There's at least one line that should be 'op.await' now.

Barring anyone else's concerns, this can then be merged.

File: src/driver/socket.rs [hunk 3]
```diff
@@ -44,6 +44,11 @@ impl Socket {
         op.await
     }
 
+    pub async fn writev<T: IoBuf>(&self, buf: Vec<T>) -> crate::BufResult<usize, Vec<T>> {
+        let op = Op::writev_at(&self.fd, buf, 0).unwrap();
+        op.writev().await
```
Comment (FrankReh) [inline src/driver/socket.rs:49]: Change this to op.await and it will compile again having merged with master.

**Comment by Noah-Kennedy** [comment]
CI is currently failing.

**Comment by FrankReh** [comment]
> CI is currently failing.

Yea. That's because last night I hit the 'auto update' button and it updated the files it could but of course not the new file that was added. And I don't think I can go into the file from here and make the trivial fix. If we don't here from the OP soon, I can fork their branch and make this PR myself.

**Comment by fasterthanlime** [comment]
> If we don't here from the OP soon, I can fork their branch and make this PR myself.

Go ahead and make your own PR! I'm not able to fix mine right now, and there's no point in blocking on me 😌

**Review by Noah-Kennedy — APPROVED** [review]

**Review by FrankReh — APPROVED** [review]

**Comment by fasterthanlime** [comment]
@FrankReh thanks for getting this past the finish line!

**Comment by FrankReh** [comment]
My pleasure. Still getting used to squash merges on GitHub and being able to push to someone else's branch for their PR but I found a how-to that made it easy. What with master changing so much now, this is going to come up often and generally the maintainers will know how to resolve differences much easier than the OP, even more so if the PR was written a while ago. Good luck with your http investigations!

**Pull request merged**